### ISSUE TITLE
Bundler platform locking causes problems out-of-box, on CI, on Heroku etc.

### ISSUE THREAD

**Issue by pond** [rubygems/rubygems#5338]
### Describe the problem as clearly as you can

I'm hopeful we can come up with a compromise to a significant pain point that's worsening over time, not getting better.

In Bundler 2.2, `PLATFORMS` in `Gemfile.lock` only includes the local machine's architecture when that machine causes a `Gemfile.lock` to be built (but confusingly, not if it is merely updated). Bundler lockfiles cease to become architecture independent, even though Ruby is supposed to be. Any attempt to install using that lockfile on any other architecture fails unless new architectures are manually added. This even includes developers upgrading e.g. macOS versions from 11 to 12 and similar, since their platform includes a Darwin kernel version. On Heroku, this broke the world for a while, as the 2.2 approach appeared to [not be backwards compatible with Bundler 2.1](https://www.moncefbelyamani.com/understanding-the-gemfile-lock-file/) that Heroku provided at the time.

Developers are forced to constantly remember to check, double check and where necessary hand-maintain their Gemfile.lock's `PLATFORMS`, to nanny things through CI (e.g. macOS development, AWS Linux CI) or just allow for the fact that they might not be on the same precise e.g. macOS version as their peers, or might be using Arch Linux, or a different CPU (e.g. M1), or whatever.

There is presently no clean way around this. Unfrozen is too permissive and has unrelated side effect of mutable Ruby gem versions, rather than _only_ dealing with native extension platforms. Bundler should not be locking itself to the arbitrary architecture of the local developer machine IMHO; its job is to lock gem versions. I am aware of [this comment](https://github.com/rubygems/rubygems/issues/4269#issuecomment-758564690) but it doesn't really seem to resonate; the "safety" argument talks about a gem which doesn't have a particular architecture variant and installs then runs on CI presumably building "generic" native extensions, CI passes, then subsequently an architecture-specific variant is pushed but it's buggy and this gets deployed _after_ the CI run to Production. If the timing of things were thus, it'd still happen regardless of what `PLATFORMS` might say - it'd still build generic on CI because no arch-specific version existed. Meanwhile, if malicious code is being pushed to RubyGems, then there are far bigger issues at play and far greater concern than specific platform variant locking.

A reasonable solution IMHO might be to provide a configuration option that allows alternative platforms, as if the running platform was included in the `PLATFORMS` array, without needing to modify `Gemfile.lock`. That could be a `bundle config` item or env var. That's not a question of just forcing to `ruby`, the resolution of which allegedly is buggy or unsafe and is the whole reason why the new 2.2 behaviour was introduced anyway - though if Bundler **guaranteed and formally documented** that forcing to `ruby` was still fully supported, perhaps that would be acceptable (now that it works again, [thanks to this fix](https://github.com/rubygems/rubygems/pull/5304)!). Otherwise we're building on sand.

### Did you try upgrading rubygems & bundler?

Yes. Bundler 2.2.x is working as designed. I do not believe this design is helpful, with widespread community posts across forums and blogs indicating the developer pain it has inadvertently introduced.

### Post steps to reproduce the problem

Cause a `Gemfile.lock` to be generated under Bundler 2.2.x on one architecture. Try to `bundle install` on another. This will fail, even though Ruby and gems all work just fine on the target architecture. In Bundler 2.1 and all earlier versions, the gems would install.

### What were you expecting to happen?

* If the gem is not compatible with the chosen platform, the native extension build will fail, just as it always has.
* If the gem can build on the chosen platform, then it should do so, and not have Bundler fail the whole thing.

### What actually happened?

Bundler refuses to install the gems:

```
Your bundle only supports platforms ["x86_64-darwin-21"] but your local platform
is x86_64-linux. Add the current platform to the lockfile with `bundle lock
--add-platform x86_64-linux` and try again.
```

**Comment by indirect** [comment]
There is no such thing as “gem compatible with the platform”. As soon as gems _have_ a platform (beyond “ruby”) they are entirely different code bases. Historically, there are extremely popular gems like Nokogiri that have completely different dependencies for the exact same version on a different platform. On a new platform, Bundler literally has to start the resolution over again from scratch, find out what gems and dependencies are needed, and create a new lockfile. Without as explicit lock for a given platform, you are asking for Bundler to silently allow new gems to be resolved and installed only in production, which is a complete nonstarter.

Can you please describe what about your workflow makes it impossible for developers to run `bundle lock` once per platform you use?

**Comment by pond** [comment]
> There is no such thing as “gem compatible with the platform”. As soon as gems have a platform (beyond “ruby”) they are entirely different code bases

But they are not different gems, don't have different APIs and don't have different major, minor or patch versions. The gem is honour-bound by contract of its API to behave the same on whatever platforms it _can_ build on.

The workflow I'd like is the one where Bundler <= 2.1 did all this just fine and nobody had to run "bundle lock" ever.

**Comment by pond** [comment]
Oh, and also:

> impossible for developers to run bundle lock once per platform you use?

Oh and - CI. It's impossible for CI to do that. The developers have to remember to 'bundle lock' not just for whatever arch their local machine is, but also include the arch that CI uses.

**Comment by whithajess** [comment]
@indirect In terms of workflow won't this always be the case? -> 

1. Developer runs `bundle install` - Generates lock file with _Local Architecture_
2. CI / CD picks up the lock file to try install the gems specified (Often specifically specifying frozen to intentionally fail and catch something that doesn't line up in the Gemfile with Gemfile.lock)
3. Fails every time unless the _CI/CD Architecture_ has been specifically added to the lockfile generated by the developer (and is maintained there)

As far as I can see the only workaround for the issue are as follows:

1. Developer runs `bundle install` - Generates lock file with  _Local Architecture_
2. Developer runs `bundle lock --add-platform X` _(knowing the remote platforms architecture)_
    * Must happen before it gets to CI/CD otherwise prevented from update due to being _frozen_
    * Must remember to do this always when _recreating_ or _initially_ creating any `Gemfile.lock` where the `remote` architecture differs

`force_ruby_platform` won't work either with  _frozen_

The only reason this isn't more widely reported is because people have started their projects on:
```
PLATFORMS
  ruby
```

And haven't recreated their `Gemfile.lock`

Everyone who starts a new project with a different _remote platform architecture_ will run into this if they use _frozen_ which is best practice.

**Comment by indirect** [comment]
@deivid-rodriguez I like that idea! I worry a little bit that this might lead to locking a lot of platforms, but I’d be interested to try it out in a prerelease.

**Comment by aprilmintacpineda** [comment]
I'm having a similar error with GIthub Actions. This used to work last time I tried it (months ago), I don't know why I have to do this now.

```
bundle install
  /Users/runner/hostedtoolcache/Ruby/3.0.2/x64/bin/bundle config --local path /Users/runner/work/beny-react-native/beny-react-native/vendor/bundle
  /Users/runner/hostedtoolcache/Ruby/3.0.2/x64/bin/bundle config --local deployment true
  Cache key: setup-ruby-bundler-cache-v3-macos-11.0-ruby-3.0.2-Gemfile.lock-3ee146a75c2f746ba7e1b2c2206de7[36](https://github.com/disruptivsolutions/beny-react-native/runs/5176473300?check_suite_focus=true#step:4:36)66e06f2b00fdcfdaaabcde931febe78d
  /Users/runner/hostedtoolcache/Ruby/3.0.2/x64/bin/bundle install --jobs 4
  Your bundle only supports platforms ["x86_64-darwin-20"] but your local platform
  is x86_64-darwin-19. Add the current platform to the lockfile with `bundle lock
  --add-platform x86_64-darwin-19` and try again.
  Took   1.12 seconds
Error: The process '/Users/runner/hostedtoolcache/Ruby/3.0.2/x64/bin/bundle' failed with exit code 16
```

- What benefit does this add?
- What if you can't control the architecture that your CI will run on? Should you then add all possible architectures that the CI is using to `Gemfile.lock`? Is that a good idea because at that point the whole thing no longer makes sense, might as well just remove it completely.

**Comment by deivid-rodriguez** [comment]
@indirect Yes, I have the same concern only it doesn't seem like a big deal in principle I think. We could, however, start by doing this only for Linux and see how it goes.

**Comment by pond** [comment]
Any further progress or more thoughts on this?

**Comment by deivid-rodriguez** [comment]
No, just that I plan to work on this soon.

**Comment by lucaspiller** [comment]
@deivid-rodriguez We are working on a platform that is developed on macOS running either Intel or Apple Silicon CPUs and Linux laptops, and run on Linux servers with Intel CPUs.

What's not clear is what problem this is trying to solve. The example you gave of nokogiri being different on different hosts/architectures makes sense, but I don't see how adding an extra flag to the `Gemfile.lock` would solve it. It seems like this is more an indication that it _should_ work, but why force developers to change their workflow? Trust people are smart and know what they are doing if they try to mix hosts/architectures.

**Comment by deivid-rodriguez** [comment]
Not sure which example you mean, but locking the specific platform in the `Gemfile.lock` file did fix platform issues as already explained. Anyways, I already admitted that this is inconvenient and that I will work on solving it, not sure what it is that you expect to get from your message.

**Comment by deivid-rodriguez** [comment]
Finally got around to working on this, feel free to give https://github.com/rubygems/rubygems/pull/5700 a try!

**Comment by pond** [comment]
@deivid-rodriguez 

Uuuh - but - I just updated to Bundler 2.3.22 and found the behaviour has become even worse! When I do `bundle update`, it's now _forcibly overwriting anything I do_ in `PLATFORMS`. I cannot put `ruby` in there, even in addition to `x86_64-darwin-21` (or the Apple Silicon equivalent) as the entire thing is thrown away and replaced.

This makes it impossible for us to move our code even between different development machines, e.g. Apple Silicon vs x86, and impossible to deploy to CI!

How can this be resolved please? This is now *an extremely critical, entirely workflow-breaking bug*. I can't even uninstall 2.3.22 to go back to the previous working version, because I'm told:

```
Gem bundler-2.3.22 cannot be uninstalled because it is a default gem
```

...so I'm basically trapped and locked into a broken version via RubyGems itself, and now unable to deploy working Ruby code _anywhere_ because - very obviously! - our deployed Staging and Production systems are not x86 Apple Macs.

**Comment by pond** [comment]
@deivid-rodriguez 

Uuuh - but - I just updated Ruby Gems and this forced me up to Bundler 2.3.22, where I found the behaviour not fixed at all - it has become even worse! When I do `bundle update`, it's now _forcibly overwriting anything I do_ in `PLATFORMS`. I cannot put `ruby` in there, even in addition to `x86_64-darwin-21` (or the Apple Silicon equivalent) as the entire thing is thrown away and replaced.

This makes it impossible for us to move our code even between different development machines, e.g. Apple Silicon vs x86, and impossible to deploy to CI!

How can this be resolved please? This is now *an extremely critical, entirely workflow-breaking bug*. I can't even uninstall 2.3.22 to go back to the previous working version, because I'm told:

```
Gem bundler-2.3.22 cannot be uninstalled because it is a default gem
```

...so I'm basically trapped and locked into a broken version via RubyGems itself, and now unable to deploy working Ruby code _anywhere_ because - very obviously! - our deployed Staging and Production systems are not x86 Apple Macs.

I note in v2.3.21:

https://github.com/rubygems/rubygems/blob/master/bundler/CHANGELOG.md#bug-fixes-1

> Fix Ruby platform incorrectly removed on `bundle update` https://github.com/rubygems/rubygems/pull/5832

...so this was either reverted in 2.3.22, or was never properly fixed in the first place.

**Comment by pond** [comment]
...as a followup, CI tells me that we were using Bundler 2.3.13, which the Bundler change log says is May 4th. The RubyGems version history shows *3*.3.13 on May 4th, so I guess that this must be the corresponding release. I had to follow https://stackoverflow.com/a/62125382 to get rid of all stray defaults, `gem uninstall bundler` to remove anything installed on top of Ruby Gems, `gem update --system 3.3.13` (consider adding `--no-doc` if you want to avoid an _extremely_ lengthy `ri` documentation installation process) and after all that (with an `rbenv rehash` for good measure!), `bundler --version` reports `2.3.13`, which was our last-known-good.

I then changed our `PLATFORMS` entry in `Gemfile.lock`, re-ran `bundle update` and under 2.3.13, this is _not_ replaced, so I'm left with a working Gemfile for local development, CI and deployments.

2.3.13 is our last-known-good, but there may be newer versions that also did not have the full-system-breaking bug I see in 2.3.22.

If there's anything you need that might help diagnose the platform-overwriting bug and get a fix sorted, please let me know and I'll get that to you ASAP, within the limits of timezones & weekends 😄

**Comment by flavorjones** [comment]
@pond I can sense that you're frustrated, and I'd like to try to help.

I've tried to reproduce what you're describing above (your most recent two comments) and I cannot with bundler v2.3.22. I'm using a Gemfile from a vanilla Rails 7 app, I'm adding platforms with `bundle lock --add-platform`, and then I change the Gemfile and `bundle update`. In this case, the `PLATFORMS` behave as expected and remain in the lockfile.

Can I ask you to open a **new issue** for this so that it may be triaged appropriately and receive some focus, without distracting from this long-running conversation? Please make sure to help the maintainers reproduce what you're seeing. Thank you!

**Comment by januszm** [comment]
It would be great if a command like `bundle platform` showed what all known platforms are to avoid unnecessary booting, and/or executing the test suite on a CI platform just to find out what it's called. This way we could pre-add e.g. `--add-platform x86_64-linux` or something else to make it work without errors. BTW. are `x86_64-linux` and `x86_64-linux-gnu` interchangeable and work the same?

**Comment by huix11** [comment]
eeee

**Comment by momolog** [comment]
> BTW. are x86_64-linux and x86_64-linux-gnu interchangeable and work the same?

I would like to understand this as well.

**Comment by deivid-rodriguez** [comment]
> It would be great if a command like bundle platform showed what all known platforms are to avoid unnecessary booting, and/or executing the test suite on a CI platform just to find out what it's called. This way we could pre-add e.g. `--add-platform x86_64-linux` or something else to make it work without errors. BTW. are `x86_64-linux` and `x86_64-linux-gnu` interchangeable and work the same?

Currently creating a new lockfile adds all potential platforms to the lockfile, and I'm working on adding those to existing lockfiles in #7896.

Regarding `x86_64-linux` and `x86_64-linux-gnu`, they should work the same, yes.

**Comment by thisismydesign** [comment]
What's extra confusing is that I'm using docker so that i don't have to worry about things like this. I would love for this to be abstracted away so that every platform is supported out of the box. Maybe get a warning on bundle install if some platforms are not supported due to missing native gems; or where it needs to fall back to a ruby platform and be potentially slower.

**Comment by deivid-rodriguez** [comment]
This is what we are doing now. Can you explain which problem you are facing?

**Comment by thisismydesign** [comment]
@deivid-rodriguez I switched to sqlite. Developing locally on Apple Silicon via Docker, also running specs on Docker in Github Actions. Got this error trying to build the Docker image:

https://github.com/thisismydesign/wealth/actions/runs/13477786684/job/37659079224
```
#19 2.953 Could not find gem 'sqlite3' with platforms 'aarch64-linux-musl', 'x86_64-linux'
#19 2.953 in rubygems repository https://rubygems.org/ or installed locally.
```

At this point I already added `x86_64-linux` platform to Gemfile.lock with `--add-platform`:
https://github.com/thisismydesign/wealth/blob/73b9912c85da747b15f39c199e9b7fe16b025caa/Gemfile.lock#L391-L393

**Comment by deivid-rodriguez** [comment]
That lockfile is using Bundler 2.4.10, that's pretty old. Try running `bundle lock --update --bundler` and see if the latest version of Bundler works more in the way you expect.

**Comment by pond** [comment]
@deivid-rodriguez FWIW, from at least about `2.5.something` and certainly in `2.6.x` this has all been working very smoothly for me on a few different deployments with quite widely varying environments. It certainly seems like a fixed issue from my perspective.

Thanks for all your work on this; as the long discussion has shown, it's a much more complex issue than I first imagined and has clearly taken significant effort to work through.
